File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/RequireFilesExistRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;

/**
 * Rule that requires specified files to exist in the project.
 */
public enum RequireFilesExistRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "requireFilesExist";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(catNode.get("parameters").get("paths")).stream()
                .filter(path -> !HarnessCheckHelper.isSafeRegularFile(root, root.resolve(path)))
                .map(path -> new Finding(severity, CATEGORY, "missing file: " + path))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/RequireFilesExistRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessPsiResults
import ai.harness.gradle.Severity
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.isSymbolicLink

/**
 * Rule that requires specified files to exist.
 */
object RequireFilesExistRule : HarnessCheckRule {
    override fun applies(manifest: JsonObject): Boolean {
        val category = "requireFilesExist"
        val catObj = manifest[category]?.jsonObject ?: return false
        val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
        return enabled
    }

    override fun validate(
        manifest: JsonObject,
        root: Path,
        psiResults: HarnessPsiResults?,
    ): Collection<Finding> {
        val category = "requireFilesExist"
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        return if (catObj == null || parametersObj == null) {
            emptyList()
        } else {
            val paths = HarnessCheck.stringArrayFrom(parametersObj, "paths")
            paths
                .filter { path ->
                    val p = root / path
                    (p.isSymbolicLink() && !HarnessCheck.isAllowedRootContractSymlink(root, p)) || !p.isRegularFile()
                }.map { path ->
                    val p = root / path
                    when {
                        p.isSymbolicLink() &&
                            !HarnessCheck.isAllowedRootContractSymlink(
                                root,
                                p,
                            )
                        -> {
                            Finding(Severity.ERROR, category, "symlink file is not allowed: $path")
                        }

                        !p.isRegularFile() -> {
                            Finding(
                                HarnessCheck.severityOf(manifest, category),
                                category,
                                "missing file: $path",
                            )
                        }

                        else -> {
                            null
                        }
                    }
                }.filterNotNull()
        }
    }
}
